### ISSUE TITLE
[REQUEST]: Documentation for RISC-V opcode

### ISSUE THREAD

**Issue by jaxvanyang** [compiler-explorer/compiler-explorer#4584]
### Is your feature request related to a problem? Please describe

Opcode documentation is an amazing feature, especially for asm learner. I am learning the RISC-V ISA, so I thought of the Compiler Explorer might help. But I find that it does not support RISC-V documentation now:

```
There was an error fetching the documentation for this opcode (Error: No documentation for 'riscv64').
```

I have been looking for a related issue for a while, but I didn't find a feature request for the documentation. I think it's wierd, so I create this issue in case of that you guys didn't plan for this.

### Describe the solution you'd like

Though I didn't find a directly related issue, I found a issue discuss about ARM documentation:
#2070

I think the solution there should be fine, the only thing needed is the documentation.
This should be helpful:
[Specifications - RISC-V International](https://riscv.org/technical/specifications/)

### Describe alternatives you've considered

Here is already a stopgap - not showing the documentation and a popup message.

### Additional context

Respect :)

**Comment by JaxVanYang** [comment]
RISC-V foundation doesn't provide a ISA documentaion which can be easily parsed. The link I attached earlier only contains PDFs, and I found it very hard to generate a opcode code from that. Luckily, somebody has been working on this stuff already: [Machine Readable Specification Data | Five EmbedDev](https://five-embeddev.com/toolchain/2021/09/10/machine-readable-spec/).

You can view a discussion about machine readable specification data there, and a lot of useful links. I suggest [five-embeddev/riscv-isa-data: Machine readable ISA data](https://github.com/five-embeddev/riscv-isa-data) for opcode doc generation.

**Comment by JaxVanYang** [comment]
Useful doc: [compiler-explorer/AddingAssemblyDocumentation.md](https://github.com/compiler-explorer/compiler-explorer/blob/main/docs/AddingAssemblyDocumentation.md)🫡

**Comment by SiyaoIsHiding** [comment]
woah good work thx a lot. I am also coding RISC-V assembly and would love to see the documentation work. I'll try to take over the rest of the work.

**Comment by jaxvanyang** [comment]
> woah good work thx a lot. I am also coding RISC-V assembly and would love to see the documentation work. Not sure whether you are still working on it. I'll try to take over the rest of the work.

Sorry for no progress, I'm busy working on my bachelor thesis. So this issue has been put off for a long time. Maybe I will work on this after graduation, but I can't promise.

I'm glad to see you are also interested in this, so the short answer is that you should try to take over.

**Comment by SiyaoIsHiding** [comment]
@jaxvanyang . May I ask why u think https://github.com/compiler-explorer/compiler-explorer/blob/main/docs/AddingAssemblyDocumentation.md can be a good source?
And meanwhile, I find this https://five-embeddev.com/quickref/instructions.html#-rv64--integer-register-immediate-instructions. Do you think it would be a better choice?

**Comment by partouf** [comment]
> @jaxvanyang . May I ask why u think https://github.com/compiler-explorer/compiler-explorer/blob/main/docs/AddingAssemblyDocumentation.md can be a good source? And meanwhile, I find this https://five-embeddev.com/quickref/instructions.html#-rv64--integer-register-immediate-instructions. Do you think it would be a better choice?

https://github.com/compiler-explorer/compiler-explorer/blob/main/docs/AddingAssemblyDocumentation.md is just the instructions to add asm documentation to CE, you can "learn" all the actual things elsewhere on the internet. Adding instruction documentation to CE is just a convenience. There's no right or wrong here.

**Comment by SiyaoIsHiding** [comment]
Oh my god I am so sorry I copy-n-pasted the wrong link!!!
I mean this one: [riscv-asm-manual/riscv-asm.md at master · riscv-non-isa/riscv-asm-manual](https://github.com/riscv-non-isa/riscv-asm-manual/blob/master/riscv-asm.md).
I wanna discuss which documentation can be a better data source considering whether it is reliable and easy to parse or not.
@jaxvanyang suggests to use [riscv-asm-manual/riscv-asm.md at master · riscv-non-isa/riscv-asm-manual](https://github.com/riscv-non-isa/riscv-asm-manual/blob/master/riscv-asm.md). But I find it hard to parse. For example, 
```txt
l{b|h|w|d} <rd>, <symbol>: load byte, half word, word or double word from global
s{b|h|w|d} <rd>, <symbol>, <rt>: store byte, half word, word or double word to global
```
Meanwhile I find another data source https://five-embeddev.com/quickref/instructions.html#-rv64--integer-register-immediate-instructions, which seems easier to parse, as all the opcodes are in tables.
Thank you for your response @partouf and so sorry for the confusion. Thank you for the guide of adding assembly documentation to CE, too. I am following it and writing scripts to add RISCV opcode docs right now :)

**Comment by jaxvanyang** [comment]
https://five-embeddev.com is a good data resources, actually I metioned it before. I suggested to use it at first, but later I added a line throught it because it's not official. Maybe I should clarify that.

> RISC-V foundation doesn't provide a ISA documentaion which can be easily parsed. The link I attached earlier only contains PDFs, and I found it very hard to generate a opcode code from that. Luckily, somebody has been working on this stuff already: [Machine Readable Specification Data | Five EmbedDev](https://five-embeddev.com/toolchain/2021/09/10/machine-readable-spec/).
>
> You can view a discussion about machine readable specification data there, and a lot of useful links. I suggest [five-embeddev/riscv-isa-data: Machine readable ISA data](https://github.com/five-embeddev/riscv-isa-data) for opcode doc generation.

[five-embeddev/riscv-isa-data: Machine readable ISA data](https://github.com/five-embeddev/riscv-isa-data)  has the data source used to generate that website. There is a big [opcodes.yaml](https://github.com/five-embeddev/riscv-isa-data/blob/master/opcodes.yaml), it should be easier to parse. I tried to parse it before, and encontered some YAML format error at first, then discontinued because it's not official. I don't know how @five-embeddev made that YAML file, maybe we could ask him for help. But after all, it is a good data source, both website and that YAML file.

For the official resource: [riscv-asm-manual/riscv-asm.md](https://github.com/riscv-non-isa/riscv-asm-manual/blob/master/riscv-asm.md), it is hard to parse so I stopped working on this issue (my bad). As for now, something works may be better than something perfect.

In summary, I think there are 3 approaches:

1. Parse the official resource: [riscv-asm-manual/riscv-asm.md](https://github.com/riscv-non-isa/riscv-asm-manual/blob/master/riscv-asm.md) or something.
2. Parse the unofficial resource, YAML in [five-embeddev/riscv-isa-data: Machine readable ISA data](https://github.com/five-embeddev/riscv-isa-data) or its website, or something.
3. Maintain a data source like [msyksphinz-self/riscv-isadoc](https://github.com/msyksphinz-self/riscv-isadoc) (or parse that repo but without privileged instructions).

Maybe we could work on the easy approach first and turn to parse the official resource when needed.

@SiyaoIsHiding

**Comment by SiyaoIsHiding** [comment]
Thanks!
Before I find out the YAML file, I wrote a script to parse the five-embeddev web page https://five-embeddev.com/quickref/instructions.html#-rv64--integer-register-immediate-instructions in this way:
```python
        tables = soup.find_all("table", class_="sortable-theme-dark")
        for table in tables:
            records = table.tbody.find_all("tr")
            for record in records:
                yield operation(record)
```
Although it is a trick based on the observation that all the opcodes are in tables with the class `sortable-theme-dark`, which may be unreliable, the script works well at least for now.
How do you think?

**Comment by jaxvanyang** [comment]
> Although it is a trick based on the observation that all the opcodes are in tables with the class `sortable-theme-dark`, which may be unreliable, the script works well at least for now.
> How do you think?

I think it's fine, as we can fix it if it broke.

**Comment by five-embeddev** [comment]
The YAML files at https://github.com/five-embeddev/riscv-isa-data/blob/master/opcodes.yaml now pass yamllint so they should be usable.

I've added an example how the tables are created here:

https://github.com/five-embeddev/riscv-isa-data/tree/master/examples/ruby

It's a stripped-down template from the quick reference.
